tests/test_paragraph_location.py

```
@@ -0,0 +1,214 @@
+    def test_out_of_range_index_raises(self, gridspan_docx):
+        with Document.open(gridspan_docx) as doc:
+            n = len(doc.list_paragraphs())
+            bad_ref = f"P{n + 99}#0000"
+            with pytest.raises(ParagraphIndexError) as exc:
+                doc.get_paragraph_location(bad_ref)
+            assert exc.value.index == n + 99
+            assert exc.value.total_paragraphs == n
+
+    def test_malformed_ref_raises_value_error(self, gridspan_docx):
+        with Document.open(gridspan_docx) as doc:
+            with pytest.raises(ValueError):
```

Ruff (0.15.13)

[warning] Use a single `with` statement with multiple contexts instead of nested `with` statements

Combine `with` statements

(SIM117)